source-google-sheets-native/source_google_sheets_native/api.py

```diff
@@ -38,8 +39,16 @@ async def fetch_rows(
         "fields": "properties,sheets.properties,sheets.data(rowData.values(effectiveFormat(numberFormat(type)),effectiveValue))",
     }
 
+    # checking if sheet is empty before model eval
+    # since the rowData field wont be returned 
+    sheet_info = json.loads(await http.request(log, url, params=params))
```

Comment:
This seems like a pretty bad way to do this check. Can `"sheets"` be zero-length? The `if len(spreadsheet.sheets) == 0` check below implies it can. Beside that, deserializing and re-serializing the sheet info looks really hacky.

What actually happens in a spreadsheet contains no rows? Is `sheet.data` `None`? We have an `assert sheet.data` a little below, so maybe we can handle the absence of data more gracefully there?

Reply:
> Can "sheets" be zero-length? The if len(spreadsheet.sheets) == 0 check below implies it can.

Based on the query we are making, the given `sheet` would need to at least exist, given that the negative return from the below query is a 400:

```python
url = f"{API}/v4/spreadsheets/{spreadsheet_id}"
params = {
    "ranges": [sheet.properties.title],
    "fields":"properties,sheets.properties,sheets.data(rowData.values(effectiveFormat(numberFormat(type)),effectiveValue))",
    }

```
We are already validating the sheet name using properties.title, so i dont know how it could return a sheet that passes that query and does not actually "exists".  It can not be validated by the `spreadsheet.sheets` under a special condition i cant really find, but we would at least receive the API payload to be able to do the data check.

> What actually happens in a spreadsheet contains no rows? Is sheet.data None? 

No. `sheet.data` does not passes model eval.  `sheet.data` Requires a specific field called rowData thats created inside the sheets["data"] field. The issue is, when the sheet has no data at all, API calls return data as:
```json
"data":  [{}]
```
So the `assert sheet.data` would never trigger that, because the code would raise at `Spreadsheet.model_validate_json` by not having the `rowData` inside `data`

> maybe we can handle the absence of data more gracefully there?

I did thought about modifing the pydantic model to allow for data to be empty and then add a check there, but i dont personally see it as "more efficient" than just receiving the payload and checking there, at the end of the day, if there is no data in that sheet, i think we should't spend a second more of processing it.
I dont have any ideas as of now, but if you have any ideas or more feedback i would love to hear it!